### PR TITLE
Support v3-lro in modular

### PR DESCRIPTION
**Pull request by MaryGao**
fixes https://github.com/Azure/autorest.typescript/issues/2166

- operation interface
```ts
export interface BudgetsCreateOrReplaceOptions extends OperationOptions {
  /** Delay to wait until next poll, in milliseconds. */
  updateIntervalInMs?: number;
}

export function createOrReplace(
  context: Client,
  name: string,
  resource: User,
  options: BudgetsCreateOrReplaceOptions = { requestOptions: {} },
): PollerLike<OperationState<User>, User>

```

- api-level helper in pollingHelpers.ts(only internal usage)
```ts
export function getLongRunningPoller<
  TResponse extends PathUncheckedResponse,
  TResult = void,
>(
  client: Client,
  processResponseBody: (result: TResponse) => PromiseLike<TResult>,
  options: GetLongRunningPollerOptions<TResponse>,
): PollerLike<OperationState<TResult>, TResult>
```

- client-level helper in restorePollerHelpers.ts

```ts
export function restorePoller<TResponse extends PathUncheckedResponse, TResult>(
  client: AzureLoadTestingContext | LoadTestRunClient,
  serializedState: string,
  sourceOperation: (
    ...args: any[]
  ) => PollerLike<OperationState<TResult>, TResult>,
  options?: RestorePollerOptions<TResult>,
): PollerLike<OperationState<TResult>, TResult> 
```

Please note RLC to adopt the new LRO is in progress so I tentatively disabled the helper generation in Modular for rest layer. And the following issues are not covered by this pr:
- https://github.com/Azure/autorest.typescript/issues/2314
- https://github.com/Azure/autorest.typescript/issues/2230
- https://github.com/Azure/autorest.typescript/issues/2313

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -177,4 +177,4 @@
       "outFiles": ["${workspaceFolder}/packages/*/dist/**/*.js"]
     }
   ]
-}
+}
```
Comment (MaryGao) [inline .vscode/launch.json:180]:
```suggestion
}

```

File: packages/typespec-ts/package.json [hunk 2]
```diff
@@ -36,7 +36,7 @@
     "regen-test-baselines": "npm run generate-tsp-only && npm run generate-tsp-only:non-branded",
     "integration-test:alone": "npm run integration-test:alone:rlc && npm run integration-test:alone:modular",
     "integration-test:alone:rlc": "cross-env TS_NODE_PROJECT=tsconfig.integration.json mocha -r ts-node/register --experimental-specifier-resolution=node --timeout 36000 ./test/integration/*.spec.ts",
-    "integration-test:alone:modular": "cross-env TS_NODE_PROJECT=tsconfig.integration.json mocha -r ts-node/register --experimental-specifier-resolution=node --timeout 36000 ./test/modularIntegration/*.spec.ts",
+    "integration-test:alone:modular": "node --loader ts-node/esm ./test/commands/prepare-deps.ts && cross-env TS_NODE_PROJECT=tsconfig.integration.json mocha -r ts-node/register --experimental-specifier-resolution=node --timeout 36000 ./test/modularIntegration/*.spec.ts",
```
Comment (MaryGao) [inline packages/typespec-ts/package.json:39]: This is temp workaround before the new LRO is not ready in RLC.

File: packages/typespec-test/test/loadtesting_modular/generated/typespec-ts/src/loadTestRun/api/operations.ts [hunk 3]
```diff
@@ -154,242 +156,22 @@ export async function _testRunDeserialize(
...
-    executedDateTime: result.body["executedDateTime"],
-    portalUrl: result.body["portalUrl"],
-    duration: result.body["duration"],
-    subnetId: result.body["subnetId"],
-    createdDateTime: result.body["createdDateTime"],
-    createdBy: result.body["createdBy"],
-    lastModifiedDateTime: result.body["lastModifiedDateTime"],
-    lastModifiedBy: result.body["lastModifiedBy"],
-  };
+  return result.body as any;
```
Comment (MaryGao) [inline packages/typespec-test/test/loadtesting_modular/generated/typespec-ts/src/loadTestRun/api/operations.ts:159]: This is temp workaround for PATCH LRO. Considering 1) there exists issues in getLroMetadata() for PATCH and 2) we don't have real case yet in DPG.

https://github.com/Azure/autorest.typescript/issues/2314

File: packages/typespec-test/test/loadtesting_modular/generated/typespec-ts/src/testRunOperations/api/operations.ts [hunk 4]
```diff
@@ -154,245 +156,22 @@ export async function _testRunDeserialize(
     throw createRestError(result);
   }
 
-  return {
```
Comment (joheredi) [inline packages/typespec-test/test/loadtesting_modular/generated/typespec-ts/src/testRunOperations/api/operations.ts:157]: Why is this deserialization thing gone with this change?

File: packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+    client = new StandardClient({
+      allowInsecureConnection: true
+    });
+  });
+
+  describe("createOrReplace", () => {
+    it("just call without await", async () => {
+      client.createOrReplace("madge", {
+        role: "contributor"
+      } as any);
```
Comment (joheredi) [inline packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts:22]: we should be asserting something here, otherwise there is no much point for the test

File: packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+    });
+    it("should await poller result directly", async () => {
+      try {
+        const result = await client.createOrReplace("madge", {
+          role: "contributor"
+        } as any);
```
Comment (joheredi) [inline packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts:28]: Why do we have to cast any?

File: packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+        assert.deepEqual(result, { name: "madge", role: "contributor" });
+      } catch (err) {
+        assert.fail(err as string);
+      }
+    });
+
+    it("should count polling counts correctly", async () => {
```
Comment (joheredi) [inline packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts:35]: I'm not sure I understand what we want to validate here?

File: packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+    });
+
+    it("await should catch inital exception", async () => {
+      try {
+        await client.createOrReplace("madge", {
+          role: "foo"
+        } as any);
+        assert.fail("Expected an exception");
+      } catch (err: any) {
+        console.log(err);
```
Comment (joheredi) [inline packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts:192]: we should remove all console lgos

File: packages/typespec-test/test/loadtesting_modular/generated/typespec-ts/src/testRunOperations/api/operations.ts (hunk 4 above)
Comment (MaryGao) [reply packages/typespec-test/test/loadtesting_modular/generated/typespec-ts/src/testRunOperations/api/operations.ts:157]: See this https://github.com/Azure/autorest.typescript/pull/2198#discussion_r1505222318.

File: packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts (hunk 8 above)
Comment (MaryGao) [reply packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts]: done.

File: packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts (hunk 7 above)
Comment (MaryGao) [reply packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts]: as we offline discussioned i would delete this case becaues it is covered by core.

File: packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+    });
+
+    it("should not override readonly attributes for operationState/result/isDone", async () => {
+      try {
+        const poller = client.createOrReplace("madge", {
+          role: "contributor"
+        } as any);
+        (poller as any).operationState = { status: "foo" };
+      } catch (err: any) {
+        assert.strictEqual(
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts]: yeah, i will remove that.

File: packages/typespec-ts/package.json [hunk 10]
```diff
@@ -71,6 +71,7 @@
     "@azure/core-paging": "^1.5.0",
     "@azure/core-lro": "^2.5.4",
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-ts/package.json:72]: no, this dependency is used for RLC cases.

File: packages/rlc-common/src/buildIndexFile.ts [hunk 11]
```diff
@@ -133,7 +133,9 @@ function generateRLCIndexForMultiClient(file: SourceFile, model: RLCModel) {
     exports.push("UnexpectedHelper");
   }
 
-  if (hasPollingOperations(model)) {
+  // TODO: Disable LRO poller tentatively and remember to remove this when new LRO in RLC is ready
+  // https://github.com/Azure/autorest.typescript/issues/2230
+  if (hasPollingOperations(model) && !model.options?.isModularLibrary) {
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/rlc-common/src/buildIndexFile.ts:138]: RLC would have two main differences 1) the interface would not be promise-like and 2) we have GAed LRO and we need to enable the flag for backward compatibility.

File: packages/rlc-common/src/buildIndexFile.ts [hunk 12]
```diff
@@ -297,7 +299,9 @@ function generateRLCIndex(file: SourceFile, model: RLCModel) {
     ]);
   }
 
-  if (hasPollingOperations(model)) {
+  // TODO: Disable LRO poller tentatively and remember to remove this when new LRO in RLC is ready
+  // https://github.com/Azure/autorest.typescript/issues/2230
+  if (hasPollingOperations(model) && !model.options?.isModularLibrary) {
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/rlc-common/src/buildIndexFile.ts:304]: > I wonder if we still need to split the cjsName and esModuleName if with Jose's esm PR merged?
I may not get the point completely, but we still need the cjs pattern for RLC in swagger way.

> I wonder if we should consider cases where we have lro operations in RLC, but no LRO operation in modular due to client.tsp's config?
The case should have no impact for this pr, because modular use its own helper to build LRO and for long term we would split RLC and modular code and they could calculate by themselves.

File: packages/typespec-test/test/healthinsights_trailmatcher_modular/generated/typespec-ts/review/health-insights-clinicalmatching.api.md [hunk 13]
```diff
@@ -0,0 +1,310 @@
+## API Report File for "@azure-rest/health-insights-clinicalmatching"
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-test/test/healthinsights_trailmatcher_modular/generated/typespec-ts/review/health-insights-clinicalmatching.api.md:1]: sure, let me remove this maybe.

File: packages/typespec-test/test/loadtesting_modular/generated/typespec-ts/src/testRunOperations/api/pollingHelpers.ts [hunk 14]
```diff
@@ -0,0 +1,120 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import {
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-test/test/loadtesting_modular/generated/typespec-ts/src/testRunOperations/api/pollingHelpers.ts:4]: the rule here is if the helper is used for client-level and operation-level. the pollingHelpers is only used for api sub folder api, but latter would be used cross API and classical layer.

File: packages/typespec-test/test/loadtesting_modular/generated/typespec-ts/src/index.ts [hunk 15]
```diff
@@ -70,6 +70,10 @@ export {
   TestRunOperationsClient,
   TestRunOperationsClientOptions,
 } from "./testRunOperations/TestRunOperationsClient.js";
+export {
+  restorePoller,
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-test/test/loadtesting_modular/generated/typespec-ts/src/index.ts:74]: we have a case in load testing where clientA has LRO but clientB has no LRO.

File: packages/typespec-test/test/openai_modular/generated/typespec-ts/src/rest/index.ts [hunk 16]
```diff
@@ -10,7 +10,6 @@ export * from "./clientDefinitions.js";
 export * from "./isUnexpected.js";
 export * from "./models.js";
 export * from "./outputModels.js";
-export * from "./pollingHelper.js";
 export {
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-test/test/openai_modular/generated/typespec-ts/src/rest/index.ts:13]: RLC interface is not ready.

File: packages/typespec-test/test/loadtesting_modular/generated/typespec-ts/src/testRunOperations/api/operations.ts [hunk 17]
```diff
@@ -154,245 +156,22 @@ export async function _testRunDeserialize(
...
-    executedDateTime: result.body["executedDateTime"],
-    portalUrl: result.body["portalUrl"],
-    duration: result.body["duration"],
-    subnetId: result.body["subnetId"],
-    createdDateTime: result.body["createdDateTime"],
-    createdBy: result.body["createdBy"],
-    lastModifiedDateTime: result.body["lastModifiedDateTime"],
-    lastModifiedBy: result.body["lastModifiedBy"],
-  };
+  return result.body as any;
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-test/test/loadtesting_modular/generated/typespec-ts/src/testRunOperations/api/operations.ts:159]: this is hack for patch lro.

File: packages/typespec-test/test/widget_dpg/spec/main.tsp [hunk 18]
```diff
@@ -112,6 +141,9 @@ result in an error.
         @statusCode code: 201;
         @body widget: Widget;
     } | WidgetError;
+ 
+    @route("/widgets/createOrReplace")
+    op createOrReplace is ResourceOperations.LongRunningResourceCreateOrReplace<User>;
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-test/test/widget_dpg/spec/main.tsp:146]: no, one for post and one for patch.

File: packages/typespec-test/test/widget_dpg/tspconfig.yaml [hunk 19]
```diff
@@ -4,6 +4,7 @@ options:
   "@azure-tools/typespec-ts":
     multiClient: false
     azureSdkForJs: false
+    generateMetadata: true
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-test/test/widget_dpg/tspconfig.yaml:7]: `generateMetadata: true` would always override.

File: packages/typespec-ts/src/modular/buildCodeModel.ts [hunk 20]
```diff
@@ -750,8 +779,8 @@ function emitLroPagingOperation(
     rlcModels,
     hierarchies
   );
-  addLroInformation(emittedOperation);
-  addPagingInformation(context.program, operation, emittedOperation);
+  addLroInformation(context, operation, emittedOperation);
+  addPagingInformation(context, operation, emittedOperation);
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-ts/src/modular/buildCodeModel.ts:783]: yeah, i prefer to split them for more clear functions. but we could refactor for common logic.

File: packages/typespec-ts/test/modularIntegration/lroRpc.spec.ts [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+      allowInsecureConnection: true
+    });
+  });
+
+  it("should await poller result directly", async () => {
+    try {
+      const result = await client.longRunningRpc({
+        prompt: "text"
+      });
+      assert.strictEqual(result.data, "text data");
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-ts/test/modularIntegration/lroRpc.spec.ts:18]: i guess restarting vscode would fix ?

File: packages/typespec-ts/test/modularIntegration/lroRpc.spec.ts [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+
+  beforeEach(() => {
+    client = new RpcClient({
+      allowInsecureConnection: true
+    });
+  });
+
+  it("should await poller result directly", async () => {
+    try {
+      const result = await client.longRunningRpc({
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-ts/test/modularIntegration/lroRpc.spec.ts:15]: i guess restarting vscode would fix ?

File: packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts [hunk 23]
```diff
@@ -0,0 +1,234 @@
...
+      } catch (err) {
+        assert.fail(err as string);
+      }
+    });
+
+    it("serialize and rehydration", async () => {
+      const poller = client.createOrReplace("madge", {
+        role: "contributor"
+      } as any);
+      const restoredPoller = await poller.serialize();
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts:34]: it would be my top priority and there are two prs one for core and one for codegen.

File: packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts [hunk 24]
```diff
@@ -0,0 +1,234 @@
...
+      const res1 = await poller;
+      const res2 = poller.result;
+      assert.deepEqual(res1, res2);
+      assert.strictEqual(operations.length, 2);
+      assert.strictEqual(states.length, 2);
+      assert.deepEqual(states, ["running", "succeeded"]);
+    });
+
+    // Skip this case: https://github.com/Azure/autorest.typescript/issues/2316
+    it.skip("should abort signal", async () => {
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-ts/test/modularIntegration/lroStardard.spec.ts:65]: i doubt that.

File: packages/typespec-ts/test/modularUnit/typeHelpers.spec.ts [hunk 25]
```diff
@@ -486,7 +488,7 @@ describe("typeHelpers", () => {
         value: "true"
       };
       const result = getType(type);
-      expect(result.name).to.equal('true');
+      expect(result.name).to.equal("true");
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-ts/test/modularUnit/typeHelpers.spec.ts:491]: my ide format them automatically.

File: packages/typespec-ts/src/modular/helpers/operationHelpers.ts [hunk 26]
```diff
@@ -107,18 +116,23 @@ export function getDeserializePrivateFunction(
...
-
   const parameters: OptionalKind<ParameterDeclarationStructure>[] = [
     {
       name: "result",
       type: getRLCResponseType(operation.rlcResponse)
     }
   ];
+  // TODO: Support LRO + paging operation
+  // https://github.com/Azure/autorest.typescript/issues/2313
+  const isLroOnly = isLROOnlyOperation(operation);
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-ts/src/modular/helpers/operationHelpers.ts:127]: > does this mean we won't do anything if it's a lro + paging now

yeah.

> why do we have other places that includes the condition of lro + paging ?

may not get the point here.

File: packages/typespec-ts/src/modular/buildLroFiles.ts [hunk 27]
```diff
@@ -0,0 +1,409 @@
...
+import { isLROOperation } from "./helpers/operationHelpers.js";
+import { ModularCodeModel, Client } from "./modularCodeModel.js";
+import path from "path";
+import { buildLroDeserDetailMap } from "./buildOperations.js";
+
+/**
+ * Always import LRO dependencies and remember to remove if unused.
+ * @param clientFile - The file to add the imports to.
+ */
+export function importLroCoreDependencies(clientFile: SourceFile) {
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-ts/src/modular/buildLroFiles.ts:12]: yeah this is the idea.

File: packages/typespec-test/test/widget_dpg/tspconfig.yaml (hunk 19 above)
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-test/test/widget_dpg/tspconfig.yaml:7]: https://github.com/Azure/autorest.typescript/issues/2363

File: packages/typespec-test/test/widget_dpg/generated/typespec-ts/sources/generated/src/index.ts [hunk 28]
```diff
@@ -5,7 +5,9 @@ export {
   WidgetServiceClient,
   WidgetServiceClientOptions,
 } from "./WidgetServiceClient.js";
+export { restorePoller, RestorePollerOptions } from "./restorePollerHelpers.js";
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-test/test/widget_dpg/generated/typespec-ts/sources/generated/src/index.ts:8]: issue created: https://github.com/Azure/autorest.typescript/issues/2364

File: packages/typespec-test/test/loadtesting_modular/generated/typespec-ts/src/testRunOperations/api/operations.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-test/test/loadtesting_modular/generated/typespec-ts/src/testRunOperations/api/operations.ts]: change the hack to keep the ser/deser logic.

File: packages/typespec-ts/src/modular/buildProjectFiles.ts [hunk 29]
```diff
@@ -82,449 +58,3 @@ export function getModuleExports(codeModel: ModularCodeModel) {
 
   return buildExportsForMultiClient(codeModel, exports);
 }
-
-// Prepare package info without scripts and devDependencies and dependencies ect
-function initPackageInfo(codeModel: ModularCodeModel) {
```
Comment (qiaozha) [inline packages/typespec-ts/src/modular/buildProjectFiles.ts:87]: why is this removed? has rlc-common already handled this?

**Review by joheredi — APPROVED** [review]

File: packages/rlc-common/src/buildIndexFile.ts [hunk 30]
```diff
@@ -297,7 +299,9 @@ function generateRLCIndex(file: SourceFile, model: RLCModel) {
     ]);
   }
 
-  if (hasPollingOperations(model)) {
+  // TODO: Disable LRO poller tentatively and remember to remove this when new LRO in RLC is ready
```
Comment (MaryGao) [inline packages/rlc-common/src/buildIndexFile.ts:302]:
```suggestion
  // TODO: Disable LRO poller tentatively
```

**Review by qiaozha — APPROVED** [review]
